File: examples/cli/bayar-gg-cli.php
#!/usr/bin/env php
<?php

require __DIR__ . '/../php/BayarGgClient.php';

$command = $argv[1] ?? 'help';
$args = parseArgs(array_slice($argv, 2));
$apiKey = getenv('BAYAR_GG_API_KEY') ?: ($args['api-key'] ?? $args['apiKey'] ?? '');
$baseUrl = getenv('BAYAR_GG_BASE_URL') ?: ($args['base-url'] ?? $args['baseUrl'] ?? 'https://www.bayar.gg/api');

if ($command !== 'help' && $apiKey === '') {
    fwrite(STDERR, "Missing API key. Set BAYAR_GG_API_KEY or pass --api-key=YOUR_API_KEY_HERE.\n\n");
    printHelp();
    exit(1);
}

$client = new BayarGgClient($apiKey, $baseUrl);

try {
    $result = runCommand($client, $command, $args);
    if ($result !== null) {
        echo json_encode($result, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE) . PHP_EOL;
    }
} catch (Throwable $e) {
    fwrite(STDERR, 'Error: ' . $e->getMessage() . PHP_EOL);
    exit(1);
}

function runCommand(BayarGgClient $client, string $command, array $args): ?array
{
    switch ($command) {
        case 'help':
            printHelp();
            return null;

        case 'methods':
            return $client->getPaymentMethods();

        case 'account-status':
            return $client->getAccountStatus();

        case 'statistics':
            return $client->getStatistics();

        case 'create-payment':
            return $client->createPayment([
                'amount' => (int)($args['amount'] ?? 10000),
                'description' => $args['description'] ?? 'Payment from BAYAR GG PHP CLI',
                'customer_name' => $args['customer-name'] ?? $args['customerName'] ?? '',
                'customer_email' => $args['customer-email'] ?? $args['customerEmail'] ?? '',
                'customer_phone' => $args['customer-phone'] ?? $args['customerPhone'] ?? '',
                'payment_method' => $args['payment-method'] ?? $args['paymentMethod'] ?? 'qris',
                'payment_url' => $args['payment-url'] ?? $args['paymentUrl'] ?? 'https://www.bayar.gg/pay',
                'callback_url' => $args['callback-url'] ?? $args['callbackUrl'] ?? '',
                'redirect_url' => $args['redirect-url'] ?? $args['redirectUrl'] ?? '',
                'file_id' => $args['file-id'] ?? $args['fileId'] ?? '',
                'content_id' => $args['content-id'] ?? $args['contentId'] ?? '',
                'product_image_id' => $args['product-image-id'] ?? $args['productImageId'] ?? '',
                'use_qris_converter' => toBool($args['use-qris-converter'] ?? $args['useQrisConverter'] ?? false),
            ]);

        case 'check-payment':
            requireArg($args, 'invoice');
            return $client->checkPayment($args['invoice']);

        case 'list-payments':
            return $client->listPayments([
                'search' => $args['search'] ?? '',
                'status' => $args['status'] ?? '',
                'payment_method' => $args['payment-method'] ?? $args['paymentMethod'] ?? '',
                'paid_via' => $args['paid-via'] ?? $args['paidVia'] ?? '',
                'start_date' => $args['start-date'] ?? $args['startDate'] ?? '',
                'end_date' => $args['end-date'] ?? $args['endDate'] ?? '',
                'page' => (int)($args['page'] ?? 1),
                'limit' => (int)($args['limit'] ?? 10),
            ]);

        case 'files':
            return $client->listFiles(toBool($args['active-only'] ?? $args['activeOnly'] ?? true));

        case 'contents':
            return $client->listContents(toBool($args['active-only'] ?? $args['activeOnly'] ?? true));

        case 'images':
            return $client->listImages(toBool($args['active-only'] ?? $args['activeOnly'] ?? true));

        case 'qris-convert':
            requireArg($args, 'qris');
            return $client->qrisConvert($args['qris'], (int)($args['nominal'] ?? $args['amount'] ?? 10000));

        case 'qris-info':
            requireArg($args, 'qris');
            return $client->qrisInfo($args['qris']);

        case 'wa-orders':
            return $client->waStoreOrders([
                'order_number' => $args['order-number'] ?? $args['orderNumber'] ?? '',
                'status' => $args['status'] ?? '',
                'search' => $args['search'] ?? '',
                'limit' => (int)($args['limit'] ?? 50),
                'offset' => (int)($args['offset'] ?? 0),
            ]);

        case 'wa-complete':
            $orderNumber = $args['order-number'] ?? $args['orderNumber'] ?? '';
            if ($orderNumber === '') {
                throw new InvalidArgumentException('Missing required option --order-number=...');
            }
            return $client->completeWaStoreOrder(
                $orderNumber,
                $args['status'] ?? 'completed',
                toBool($args['notify'] ?? true)
            );

        case 'merchant-status':
            return $client->merchantStatus();

        case 'merchant-info':
            requireArg($args, 'provider');
            return $client->merchantInfo($args['provider']);

        case 'merchant-balance':
            requireArg($args, 'provider');
            return $client->merchantBalance($args['provider']);

        case 'merchant-history':
            requireArg($args, 'provider');
            return $client->merchantHistory($args['provider'], (int)($args['limit'] ?? 20));

        case 'merchant-set-qris':
            requireArg($args, 'provider');
            return $client->merchantSetQris($args['provider'], (string)($args['qris-string'] ?? $args['qrisString'] ?? ''));

        case 'merchant-disconnect':
            requireArg($args, 'provider');
            return $client->merchantDisconnect($args['provider']);

        case 'merchant-connect':
            requireArg($args, 'provider');
            requireArg($args, 'action');
            return $client->merchantConnect(buildMerchantBody($args));

        default:
            printHelp();
            throw new InvalidArgumentException('Unknown command: ' . $command);
    }
}

function buildMerchantBody(array $args): array
{
    // Petakan opsi CLI → field body accounts-connect (mendukung semua alur connect).
    $map = [
        'provider' => 'provider', 'action' => 'action', 'host' => 'host',
        'username' => 'username', 'password' => 'password', 'mid' => 'mid', 'tid' => 'tid',
        'phone' => 'phone', 'otp' => 'otp', 'pin' => 'pin', 'method' => 'method',
        'connect-token' => 'connect_token', 'outlet-id' => 'outlet_id',
        'account-id' => 'account_id', 'qris-string' => 'qris_string',
    ];
    $body = [];
    foreach ($map as $argKey => $bodyKey) {
        if (isset($args[$argKey]) && $args[$argKey] !== true && $args[$argKey] !== '') {
            $body[$bodyKey] = $args[$argKey];
        }
    }
    return $body;
}

function parseArgs(array $rawArgs): array
{
    $parsed = [];
    foreach ($rawArgs as $arg) {
        if (strpos($arg, '--') !== 0) {
            continue;
        }
        $arg = substr($arg, 2);
        $parts = explode('=', $arg, 2);
        $key = $parts[0];
        $value = $parts[1] ?? true;
        $parsed[$key] = $value;
        $parsed[toCamel($key)] = $value;
    }
    return $parsed;
}

function toCamel(string $key): string
{
    return preg_replace_callback('/-([a-z])/', fn($m) => strtoupper($m[1]), $key);
}

function requireArg(array $args, string $key): void
{
    if (empty($args[$key])) {
        throw new InvalidArgumentException("Missing required option --{$key}=...");
    }
}

function toBool($value): bool
{
    if (is_bool($value)) return $value;
    return in_array(strtolower((string)$value), ['1', 'true', 'yes', 'on'], true);
}

function printHelp(): void
{
    echo <<<TXT
BAYAR GG API PHP CLI

Usage:
  BAYAR_GG_API_KEY=YOUR_API_KEY_HERE php examples/cli/bayar-gg-cli.php <command> [options]

Commands:
  methods                         List payment methods
  account-status                  Get account status
  statistics                      Get payment statistics
  create-payment                  Create payment link
  check-payment --invoice=...     Check payment status
  list-payments                   List payments
  files                           List digital files
  contents                        List hidden contents
  images                          List product images
  qris-convert --qris=...         Convert static QRIS
  qris-info --qris=...            Decode QRIS info
  wa-orders                       List WhatsApp Store orders
  wa-complete --order-number=...  Complete WhatsApp Store order

Merchant API (paket Premium "Semua Fitur"):
  merchant-status                       Status koneksi OVO/BRI/GoPay/Livin
  merchant-info --provider=...          Info akun merchant
  merchant-balance --provider=...       Saldo merchant (ovo/gopay/livin)
  merchant-history --provider=... --limit=20   Riwayat transaksi
  merchant-set-qris --provider=bri|livin --qris-string=...   Set/hapus QRIS statis
  merchant-disconnect --provider=...    Putuskan akun
  merchant-connect --provider=... --action=...  Connect step (lihat API Docs)
      contoh BRI:   merchant-connect --provider=bri --action=connect --host=... --username=... --password=... --mid=... --tid=...
      contoh OVO:   merchant-connect --provider=ovo --action=otp_send --phone=08...
                    merchant-connect --provider=ovo --action=otp_verify --connect-token=cs_xxx --otp=123456
                    merchant-connect --provider=ovo --action=pin --connect-token=cs_xxx --pin=123456

Examples:
  php examples/cli/bayar-gg-cli.php methods
  php examples/cli/bayar-gg-cli.php create-payment --amount=10000 --description="Order #1001" --payment-method=qris --payment-url=https://www.bayar.gg/pay
  php examples/cli/bayar-gg-cli.php check-payment --invoice=PAY-USERNAME-000001

TXT;
}

File: examples/cli/bayar_gg_cli.py
#!/usr/bin/env python3

from __future__ import annotations

import argparse
import json
import os
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1] / "python"))

from bayar_gg_client import BayarGgClient


def main() -> int:
    parser = build_parser()
    args = parser.parse_args()

    api_key = args.api_key or os.getenv("BAYAR_GG_API_KEY", "")
    base_url = args.base_url or os.getenv("BAYAR_GG_BASE_URL", "https://www.bayar.gg/api")

    if not api_key:
        parser.error("Missing API key. Set BAYAR_GG_API_KEY or pass --api-key.")

    client = BayarGgClient(api_key, base_url)

    try:
        result = run_command(client, args)
    except Exception as exc:
        print(f"Error: {exc}", file=sys.stderr)
        return 1

    print(json.dumps(result, indent=2, ensure_ascii=False))
    return 0


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description="BAYAR GG API Python CLI")
    parser.add_argument("--api-key", default="", help="BAYAR GG API key")
    parser.add_argument("--base-url", default="", help="BAYAR GG API base URL")

    sub = parser.add_subparsers(dest="command", required=True)

    sub.add_parser("methods", help="List payment methods")
    sub.add_parser("account-status", help="Get account status")
    sub.add_parser("statistics", help="Get payment statistics")

    create = sub.add_parser("create-payment", help="Create payment link")
    create.add_argument("--amount", type=int, default=10000)
    create.add_argument("--description", default="Payment from BAYAR GG Python CLI")
    create.add_argument("--customer-name", default="")
    create.add_argument("--customer-email", default="")
    create.add_argument("--customer-phone", default="")
    create.add_argument("--payment-method", default="qris")
    create.add_argument("--callback-url", default="")
    create.add_argument("--redirect-url", default="")
    create.add_argument("--file-id", default="")
    create.add_argument("--content-id", default="")
    create.add_argument("--product-image-id", default="")
    create.add_argument("--use-qris-converter", action="store_true")

    check = sub.add_parser("check-payment", help="Check payment status")
    check.add_argument("--invoice", required=True)

    list_payments = sub.add_parser("list-payments", help="List payments")
    list_payments.add_argument("--search", default="")
    list_payments.add_argument("--status", default="")
    list_payments.add_argument("--payment-method", default="")
    list_payments.add_argument("--paid-via", default="")
    list_payments.add_argument("--start-date", default="")
    list_payments.add_argument("--end-date", default="")
    list_payments.add_argument("--page", type=int, default=1)
    list_payments.add_argument("--limit", type=int, default=10)

    for name in ("files", "contents", "images"):
        command = sub.add_parser(name, help=f"List {name}")
        command.add_argument("--include-inactive", action="store_true")

    qris_convert = sub.add_parser("qris-convert", help="Convert static QRIS")
    qris_convert.add_argument("--qris", required=True)
    qris_convert.add_argument("--nominal", type=int, default=10000)

    qris_info = sub.add_parser("qris-info", help="Decode QRIS info")
    qris_info.add_argument("--qris", required=True)

    wa_orders = sub.add_parser("wa-orders", help="List WhatsApp Store orders")
    wa_orders.add_argument("--order-number", default="")
    wa_orders.add_argument("--status", default="")
    wa_orders.add_argument("--search", default="")
    wa_orders.add_argument("--limit", type=int, default=50)
    wa_orders.add_argument("--offset", type=int, default=0)

    wa_complete = sub.add_parser("wa-complete", help="Complete WhatsApp Store order")
    wa_complete.add_argument("--order-number", required=True)
    wa_complete.add_argument("--status", default="completed")
    wa_complete.add_argument("--no-notify", action="store_true")

    return parser


def run_command(client: BayarGgClient, args: argparse.Namespace) -> dict:
    if args.command == "methods":
        return client.get_payment_methods()
    if args.command == "account-status":
        return client.get_account_status()
    if args.command == "statistics":
        return client.get_statistics()
    if args.command == "create-payment":
        return client.create_payment({
            "amount": args.amount,
            "description": args.description,
            "customer_name": args.customer_name,
            "customer_email": args.customer_email,
            "customer_phone": args.customer_phone,
            "payment_method": args.payment_method,
            "callback_url": args.callback_url,
            "redirect_url": args.redirect_url,
            "file_id": args.file_id,
            "content_id": args.content_id,
            "product_image_id": args.product_image_id,
            "use_qris_converter": args.use_qris_converter,
        })
    if args.command == "check-payment":
        return client.check_payment(args.invoice)
    if args.command == "list-payments":
        return client.list_payments(
            search=args.search,
            status=args.status,
            payment_method=args.payment_method,
            paid_via=args.paid_via,
            start_date=args.start_date,
            end_date=args.end_date,
            page=args.page,
            limit=args.limit,
        )
    if args.command == "files":
        return client.list_files(active_only=not args.include_inactive)
    if args.command == "contents":
        return client.list_contents(active_only=not args.include_inactive)
    if args.command == "images":
        return client.list_images(active_only=not args.include_inactive)
    if args.command == "qris-convert":
        return client.qris_convert(args.qris, args.nominal)
    if args.command == "qris-info":
        return client.qris_info(args.qris)
    if args.command == "wa-orders":
        return client.wa_store_orders(
            order_number=args.order_number,
            status=args.status,
            search=args.search,
            limit=args.limit,
            offset=args.offset,
        )
    if args.command == "wa-complete":
        return client.complete_wa_store_order(args.order_number, args.status, notify=not args.no_notify)

    raise ValueError(f"Unknown command: {args.command}")


if __name__ == "__main__":
    raise SystemExit(main())
